### ISSUE TITLE
DependencyWarning is somewhat misused

### ISSUE THREAD

**Issue by mwichmann** [SConsProject/scons#4461]
SCons scanners often emit a `DependencyWarning` if an "include file" detected during scanning is not found.  This usage should, at least in some cases, be in some sort of "informational" category, rather than "warning". The C/C++ scanner, for example, wants to follow the policy that system headers (`<stdio.h>`, etc.) not be tracked as implicit dependencies. It implements that by searching for a scanned include using the paths in `CPPPATHS`. Any includes scanned from files which are not found are added to the scanner object's `missing` attribute, which is later used to issue `DependencyWarning`.  Not-found files just aren't added as deps, which is how the system-header thing works - this isn't exactly right, but it works fine: when you actually compile, any missing headers will cause a compilation error anyway, as you expect, so it's not our job to fail the build because they're missing.  However, a warning is something that's potentially a problem, and for the system headers, it's not a problem.

We just had a Discord discussion with a developer who wanted to have warnings cause failures, which is a moderately common thing for people to do when being strict about their programming environment. That caused the build to abort, even though the failure to find `stdio.h` in `CPPPATHS` is not indicative of a problem at all:

```py
SCons.Warnings.warningAsException(True)
SetOption('warn', 'all')
```

If we're going to consider that to be valid usage, then the system headers problem needs a different solution, or perhaps more simply, it shouldn't be a warning.  I understand the desire to be able to let you know about things which were scanned as potential dependencies, but which were not actually added as dependencies - that's a useful debugging tool for non-system headers. We don't *know* something is a system header, we're just triggering off not-found and assuming it's probably a system header. Perhaps another approach would be to present missing-deps under a `--debug` option rather than through a warning?

**Comment by mwichmann** [comment]
Added note: I'd really like to see this intentional exclusion ("it's probably a system header, so it's okay we didn't find it in our limited search") documented, in code comments if nowhere else.  I had to guess my way to why the Fortran scanner was excluding deps, and ended up also excluding not-found module files, which are a different category, as they can be generated, and so (probably) should have dependencies generated for them.

**Comment by gravityfox** [comment]
This might deserve its own issue, but the fact that the Fortran scanner doesn't generate a dependency when it can't locate a (generatable) module file does cause problems. Specifically running the build can fail depending on the compilation order, but a subsequent build can then succeed (when using multiple jobs) once the missing modules have been generated.

I'm basically a passerby in this instance with some Fortran code I was asked to test, which is only 2 modules, so it's not a problem for me, but in a project with deep module dependencies I foresee this being a much more annoying problem.

**Comment by mwichmann** [comment]
Yes, you're right.  I believe Fortran doesn't use the same concept of "system headers" that C and C++ do - can you confirm? The "trick" described above exists to avoid building dependencies on those, which tend to change very slowly and really are external to the project itself.

In any case, getting generated files into the dependency tree resolves the issue, we just have to get that fix landed somehow. PR #4452 stalled out, I've been thinking of resurrecting it.

**Comment by gravityfox** [comment]
I am not well-versed in Fortran, but a cursory search indicates that an unqualified `use` statement can refer to either a user module or intrinsic module, where the compiler attempts to search for the user module first, with a fallback to the intrinsic module.

I think a mechanism that can resolve this issue is a type of node dependency that is "best-effort", in the sense that it gets constructed, but if that dependency cannot be found or built, it is discarded without error. There is another, related case that also requires this line of thinking, which is generated header files from Lex/Yacc, although in practice I have yet to encounter a build order that triggers this potential issue. (There are other complications with CXXFile, variant directories, and header resolution that I've noticed as well, which does influence this case in my project. The gist of it is that the generated sources need to be placed into the real source directory even when variant directories is in effect to avoid compiler header resolution issues.)

I don't know how difficult it would be to lookup from the list of buildable targets to see if an implicit dependency can be satisfied that way, but that might be a reasonable way to go about handling cases like this. I suspect that in the Fortran case this needs to be a 2-phase process since you need to construct dependencies for source files before you know what targets those source files build. This is just my own speculation though. I am still pretty new to SCons, and I have never worked on build system internals before, so I don't know what the fundamental hurdles are.

**Comment by mwichmann** [comment]
I have worked on lex/yacc dependencies also, so if there are still issues there it would be good to hear about them (there's the pending issue #4206 which is somewhat different).

Going back and discarding isn't something that happens now... if a dependency is identified, and not skipped, a node is made for it and it's made part of the graph, and stays there. It's an interesting thought, but I don't personally have the intimate knowledge of that subsystem to know whether such a change is workable (or even desirable).

**Comment by bdbaddog** [comment]
@gravityfox - the scanner phase happens after the emitter phase, so assuming the lex/yacc is informing scons about the generated file, the scanner should be finding it.  Can you create a simple reproducer of your issue and create a new github issue with it?
